New Transmission component

MatteGary opened the pull request:
New Transmission component and related sensor.

## Description:
With this Transmission component Home Assistant is capable of sending event in case of a torrent is completed or started. The previous behavior is mantained with an updated sensor. Now it shows also number of Started or Completed torrent.
 
**Documentation:** 
**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io):** home-assistant/home-assistant.io#7816

## Breaking Changes
With the introduction of the Transmission component, the configuration for Transmission sensor and switch has been moved to the component configuration. Configuration must be changed accordingly.

## Example entry for `configuration.yaml`:
```
transmission:
  host: 192.168.1.1
  username: user
  password: psw
  port: 8181
  monitored_variables:
  - current_status
  - completed_torrents
  - started_torrents

```
 ## Example entry for `automation.yaml`:
```  
- alias: Completed Torrent
    trigger:
      platform: event
      event_type: transmission_downloaded_torrent
    action:
      service: notify.telegram_notifier
      data_template:
        title: "Torrent completed!"
        message: "{{trigger.event.data.name}}"
```

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

MatteGary committed:
New Transmission component and interaction

First commit for New Transmission component and interaction

MatteGary committed:
Fix commit

MatteGary committed:
Fix commit

fabaff commented:
There is a `transmission` switch platform which needs to be updated as well.

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,171 @@
+    vol.Required(CONF_USERNAME): cv.string,
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_MONITORED_VARIABLES, default=['current_status']):
```
We prefer to use `CONF_MONITORED_CONDITIONS` constant.

MartinHjelmare commented on homeassistant/components/sensor/transmission.py:
```diff
@@ -32,13 +31,15 @@
 
 SCAN_INTERVAL = timedelta(minutes=2)
 
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
```
The platform schema should be removed if we load the platform via discovery from the component.

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,171 @@
+        tmp_started_torrents = list(
+            set(actual_started_torrents).difference(
+                self.started_torrents))
+        if tmp_started_torrents:
```
This can be removed.

MatteGary replied:
Removed how? I need the tmp_started_torrents variables for event triggering

MartinHjelmare replied:
We don't need to check if it's non empty before iterating the list.

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,171 @@
+            self.checkCompletedTorrent()
+            self.checkStartedTorrent()
+
+            _LOGGER.info("Torrent Data updated")
```
Debug max.

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,171 @@
+                self.completed_torrents))
+        if tmp_completed_torrents:
+            for var in tmp_completed_torrents:
+                _LOGGER.info("Completed Torrent found")
```
Debug max.

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,171 @@
+        'client_name': config[DOMAIN].get(CONF_NAME)}
+    load_platform(hass, 'sensor', DOMAIN, sensorconfig)
+
+    _LOGGER.info("Transmission component setup completed.")
```
This can be removed. It's already logged by the core.

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,171 @@
+"""
+from datetime import timedelta
+import logging
+import voluptuous as vol
```
Please add a blank line between standard library, 3rd party and homeassistant import groups.

OttoWinter replied:
https://developers.home-assistant.io/docs/en/development_guidelines.html#ordering-of-imports

MartinHjelmare replied:
Bump.

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,171 @@
+        from transmissionrpc.error import TransmissionError
+        try:
+            """Initialize the Transmission RPC API"""
+            host = config[DOMAIN].get(CONF_HOST)
```
Use `dict[key]` for required config keys and keys with default config schema values.

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,171 @@
+        import transmissionrpc
+        from transmissionrpc.error import TransmissionError
+        try:
+            """Initialize the Transmission RPC API"""
```
The docstring should be just after the function head.

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,171 @@
+            api = transmissionrpc.Client(
+                host, port=port, user=username, password=password)
+
+            """Initialize the Transmission data object."""
```
Remove this.

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,171 @@
+            self._api = api
+            self.completed_torrents = []
+            self.started_torrents = []
+            self.initTorrentList()
```
It would be good to move all the side effects that do I/O out of init method.

MatteGary replied:
You mean this lines?
```
self.completed_torrents = []
self.started_torrents = []
```
Should i put it outside of the init right after the class definition?

MartinHjelmare replied:
In github code review, it's always the bottom most line of the shown code box, that the comment is directed at. In this case line 94.

Can we call `initTorrentList` after init from `setup` function instead?

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,171 @@
+        tmp_completed_torrents = list(
+            set(actual_completed_torrents).difference(
+                self.completed_torrents))
+        if tmp_completed_torrents:
```
Remove this.

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,171 @@
+                        'name': var})
+        self.started_torrents = actual_started_torrents
+
+    def getStartedTorrentCount(self):
```
Function names should be lowercase snakecase.

MartinHjelmare commented on homeassistant/components/sensor/transmission.py:
```diff
@@ -49,39 +50,22 @@
-        _LOGGER.warning(
-            "Unable to connect to Transmission client: %s:%s", host, port)
-        raise PlatformNotReady
+
```
Add a guard clause here that checks if discovery_info is None and return if so.

MartinHjelmare commented on homeassistant/components/sensor/transmission.py:
```diff
@@ -49,39 +50,22 @@
     """Representation of a Transmission sensor."""
 
-    def __init__(self, sensor_type, transmission_api, client_name):
+    def __init__(self, sensor_type, transmission_api, client_name, hass):
```
Don't change the entity signature.

MartinHjelmare commented on homeassistant/components/sensor/transmission.py:
```diff
@@ -90,6 +74,8 @@ def __init__(self, sensor_type, transmission_api, client_name):
         self.client_name = client_name
         self.type = sensor_type
+        self.started_torrents = []
+        self.first_run = True
```
Why do we need this?

MatteGary committed:
Fix + Switch checkin

Fix according to failed build and request, first checkin for Turtle Mode Switch in Transmission, still have to figure it out why it's not working.

MatteGary committed:
Bugfixing

fabaff commented on homeassistant/components/sensor/transmission.py:
```diff
@@ -10,20 +10,16 @@
-from homeassistant.const import (
-    CONF_HOST, CONF_MONITORED_VARIABLES, CONF_NAME, CONF_PASSWORD, CONF_PORT,
-    CONF_USERNAME, STATE_IDLE)
+from homeassistant.const import (STATE_IDLE)
```
Parentheses can be removed.

fabaff commented on homeassistant/components/sensor/transmission.py:
```diff
@@ -10,20 +10,16 @@
 
-REQUIREMENTS = ['transmissionrpc==0.11']
+DEPENDENCIES = ['transmission']
+DATA_TRANSMISSION = 'TRANSMISSION'
```
Import this from the component.

fabaff commented on homeassistant/components/sensor/transmission.py:
```diff
@@ -32,47 +28,22 @@
-
-        _LOGGER.warning(
-            "Unable to connect to Transmission client: %s:%s", host, port)
+    if (discovery_info is None):
```
Parentheses can be removed.

fabaff commented on homeassistant/components/sensor/transmission.py:
```diff
@@ -32,47 +28,22 @@
-        _LOGGER.warning(
-            "Unable to connect to Transmission client: %s:%s", host, port)
+    if (discovery_info is None):
+        _LOGGER.warning("Unable to connect to Transmission client.")
```
Log messages should not have a period at the end.

fabaff commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,189 @@
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'transmission'
+DATA_TRANSMISSION = 'TRANSMISSION'
```
```suggestion
DATA_TRANSMISSION = 'data_transmission'
```

fabaff commented on homeassistant/components/sensor/transmission.py:
```diff
@@ -10,20 +10,16 @@
 DEFAULT_NAME = 'Transmission'
-DEFAULT_PORT = 9091
 
 SENSOR_TYPES = {
```
This dict is already present in the component.

fabaff commented on homeassistant/components/switch/transmission.py:
```diff
@@ -9,52 +9,32 @@
+    dev = []
+    dev.append(TransmissionSwitch(transmission_api, name))
 
+    add_entities(dev, True)
```
Can be shorter.

```suggestion
    add_entities([TransmissionSwitch(transmission_api, name)], True)
```

fabaff commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,189 @@
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_USERNAME): cv.string,
```
Username and password would still be optional as it's not a requirement of the Transmission client to set them.

fabaff commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,189 @@
+SCAN_INTERVAL = timedelta(minutes=2)
+
+
+def setup(hass, config):
```
Docstring is missing.

fabaff commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,189 @@
+    hass.data[DATA_TRANSMISSION] = TransmissionData(hass, config)
+    hass.data[DATA_TRANSMISSION].init_torrent_list()
+
+    def refresh(event_time):
```
Docstring is missing.

fabaff commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,189 @@
+                    'name': var})
+        self.started_torrents = actual_started_torrents
+
+    def get_started_torrent_count(self):
```
Docstring is missing.

fabaff commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,189 @@
+    def get_started_torrent_count(self):
+        return len(self.started_torrents)
+
+    def get_completed_torrent_count(self):
```
Docstring is missing.

fabaff commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,189 @@
+    def get_completed_torrent_count(self):
+        return len(self.completed_torrents)
+
+    def set_alt_speed_enabled(self, is_enabled):
```
Docstring is missing.

fabaff commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,189 @@
+    def set_alt_speed_enabled(self, is_enabled):
+        self._api.set_session(alt_speed_enabled=is_enabled)
+
+    def get_alt_speed_enabled(self):
```
Docstring is missing.

fabaff commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,189 @@
+        return self.get_session().alt_speed_enabled
+
+    @Throttle(SCAN_INTERVAL)
+    def get_session(self):
```
Docstring is missing.

fabaff commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,189 @@
+        self.completed_torrents = actual_completed_torrents
+
+    def check_started_torrent(self):
+        """Get started torrent functionality"""
```
Docstring should end with a period.

MatteGary committed:
Fix commit

Multiple fix

MatteGary committed:
Fix

MatteGary commented:
Hi, how can I fix the compilation error

`homeassistant/components/transmission.py:82:8: E1120: No value for argument 'hass_config' in function call (no-value-for-parameter)`

?

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,199 @@
+    discovery.load_platform(hass, 'sensor', DOMAIN, sensorconfig)
+
+    if config[DOMAIN][TURTLE_MODE] is True:
+        discovery.load_platform(hass, 'switch', DOMAIN, sensorconfig)
```
Pass the original `config` as fifth argument.

MatteGary committed:
fix for missing config

MatteGary commented:
Travis is keep failing, how can I fix this error?

`InvocationError for command '/home/travis/build/home-assistant/home-assistant/.tox/lint/bin/python script/gen_requirements_all.py validate' (exited with code 1)`

MartinHjelmare commented:
Please run `script/gen_requirements_all.py` and commit the changes.

MatteGary committed:
Update on requirements_all.txt

MatteGary committed:
Fix in requirements_all.txt

homeassistant commented:
Hello @MatteGary,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`f8b0fea7387df4fa71236f10cfe08295231078d7`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;19230&#x2F;commits&#x2F;f8b0fea7387df4fa71236f10cfe08295231078d7) - This commit has something that looks like an email address (matteogaribotti@imac-di-matteo.local). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

MatteGary committed:
Fix

MatteGary committed:
Fix for build

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,200 @@
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+SCAN_INTERVAL = timedelta(minutes=2)
```
Also import this to the platform(s). `SCAN_INTERVAL` should be used in platforms to set the default update interval of polling entities.

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,200 @@
+
+def setup(hass, config):
+    """Set up the Transmission Component."""
+    hass.data[DATA_TRANSMISSION] = TransmissionData(hass, config)
```
Save the data instance also in a local variable that we can use here in setup.
```py
tm_data = hass.data[DATA_TRANSMISSION] = TransmissionData(hass, config)
```

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,200 @@
+    track_time_interval(hass, refresh, SCAN_INTERVAL)
+
+    sensorconfig = {
+        'sensor_types': SENSOR_TYPES,
```
We don't need to pass constants in discovery_info. Just import those from the component to the platforms.

MatteGary replied:
How can I do that? If I try 
`from homeassistant.components import transmission`

I get this error
`ImportError: cannot import name 'transmission'
Make sure all dependencies are installed`

MartinHjelmare replied:
Since we depend on the component in the platforms, it should work. What is giving you that error?

MatteGary replied:
I copied and pasted your suggestion, but it still give me error while starting up HA, full error logged is 

`Error loading custom_components.sensor.transmission. Make sure all dependencies are installed
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.6/site-packages/homeassistant/loader.py", line 92, in get_component
    module = importlib.import_module(path)
  File "/usr/local/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/homeassistant/.homeassistant/custom_components/sensor/transmission.py", line 9, in <module>
    from homeassistant.components.transmission import (
ModuleNotFoundError: No module named 'homeassistant.components.transmission'`

MartinHjelmare replied:
You're using a custom component. That won't work. Set up a development environment and use the paths that will be submitted in the PR.

https://developers.home-assistant.io/docs/en/development_environment.html

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,200 @@
+        'sensor_types': SENSOR_TYPES,
+        'sensors': config[DOMAIN][CONF_MONITORED_CONDITIONS],
+        'client_name': config[DOMAIN][CONF_NAME],
+        'component_name': DATA_TRANSMISSION}
```
See above.

MartinHjelmare replied:
Bump.

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,200 @@
+
+            _LOGGER.warning(
+                "Unable to connect to Transmission client: %s:%s", host, port)
+            raise PlatformNotReady
```
This only works in platform setup. Remove it here.

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,200 @@
+            self.started_torrents = []
+            self.hass = hass
+
+        except TransmissionError as error:
```
We don't want side effects in init method. Please create the api first, then pass that to the transmission data instance when creating that. If there's an error we should `return False` from `setup` before creating the transmission data instance.

MatteGary replied:
I'm not sure of what you mean... What should I change in the init method?

MartinHjelmare replied:
See below.

MartinHjelmare commented on homeassistant/components/sensor/transmission.py:
```diff
@@ -4,89 +4,58 @@
-        _LOGGER.warning(
-            "Unable to connect to Transmission client: %s:%s", host, port)
+    if discovery_info is None:
+        _LOGGER.warning("Unable to connect to Transmission client")
```
Just `return` here. Don't log warning or raise error.

MartinHjelmare commented on homeassistant/components/switch/transmission.py:
```diff
@@ -6,54 +6,29 @@
-    password = config.get(CONF_PASSWORD)
-    port = config.get(CONF_PORT)
+    if discovery_info is None:
+        _LOGGING.warning("Unable to connect to Transmission client.")
```
See above.

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,200 @@
+            password = config[DOMAIN][CONF_PASSWORD]
+            port = config[DOMAIN][CONF_PORT]
+
+            api = transmissionrpc.Client(
```
Move creating the api to `setup`. Pass `api` to `TransmissionData` when creating that.

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,200 @@
+
+            api = transmissionrpc.Client(
+                host, port=port, user=username, password=password)
+            api.session_stats()
```
Also move this to `setup`.

MatteGary committed:
fix

MatteGary committed:
Fix

MatteGary committed:
Fix (again)

MartinHjelmare commented on homeassistant/components/sensor/transmission.py:
```diff
@@ -4,89 +4,59 @@
+    transmission_api = hass.data[component_name]
+    monitored_variables = discovery_info['sensors']
+    name = discovery_info['client_name']
+    sensor_types = SENSOR_TYPES
```
We don't need to store the constant in a local variable. Use the constant directly.

MartinHjelmare commented on homeassistant/components/sensor/transmission.py:
```diff
@@ -4,89 +4,59 @@
+    if discovery_info is None:
+        return
+
+    component_name = DATA_TRANSMISSION
```
Use the constant directly.

MartinHjelmare commented on homeassistant/components/sensor/transmission.py:
```diff
@@ -4,89 +4,59 @@
     dev = []
-    for variable in config[CONF_MONITORED_VARIABLES]:
-        dev.append(TransmissionSensor(variable, transmission_api, name))
+    for variable in monitored_variables:
```
Call the iterable `sensor_type`. `variable` is too generic.

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,193 @@
+            host, port=port, user=username, password=password)
+        api.session_stats()
+    except TransmissionError as error:
+        if str(error).find("401: Unauthorized"):
```
Is it ok to continue after all other errors caught here?

It would be best to make the distinction in the library, by subclassing a base library specific error.

MartinHjelmare replied:
This question hasn't been answered.

MatteGary replied:
What do you mean with "continue after all other errors"?

If you mean that TransmissionError is a too generic kind of error, and so it may catch other kind of error, you're right. But in any error case the setup of the platform should stop. Since I'm not the owner of the transmissionrpc package I can't make modification to implement other kind of more specific error.

MartinHjelmare replied:
I'm assuming there are other error types since you're checking for a string in the error argument, and only upon match we treat it as something to act on.

Checking for a string seems fragile. But I guess it's ok, if there's no better way right now.

MartinHjelmare replied:
I see now that it's only the log message that's indented under the if check.

MatteGary replied:
Yeah I was about to comment on that. It’s the previous behavior, I agree that it should be improved in order to log the error message in any case... maybe in another PR :D

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,193 @@
+        'component_name': DATA_TRANSMISSION}
+    discovery.load_platform(hass, 'sensor', DOMAIN, sensorconfig, config)
+
+    if config[DOMAIN][TURTLE_MODE] is True:
```
```py
if config[DOMAIN][TURTLE_MODE]:
```

MartinHjelmare commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,193 @@
+
+    def get_alt_speed_enabled(self):
+        """Get the alternative speed flag."""
+        return self.get_session().alt_speed_enabled
```
We shouldn't have two different interfaces that updates state. Update state in `update` and save that in a cache, ie instance attributes. Then get the state from the cache in the entitites. Then we can remove the throttles.

MatteGary committed:
Fix

MatteGary committed:
Fix indentation

MatteGary committed:
Fix indentation

MatteGary committed:
Fix Throttle

MatteGary committed:
Update .coveragerc

fabaff commented on .coveragerc:
```diff
@@ -348,6 +348,9 @@ omit =
     homeassistant/components/*/tradfri.py
+	
+    homeassistant/components/transmission.py
+    homeassistant/components/*/transmission.py
```
The sensor and the switch are already listed. They need to be removed.

fabaff commented on homeassistant/components/transmission.py:
```diff
@@ -0,0 +1,188 @@
+    CONF_USERNAME,
+    CONF_MONITORED_CONDITIONS,
+)
+import homeassistant.helpers.config_validation as cv
```
If you don't want to do the ordering manually then use `isort`.

MatteGary committed:
Fix import and coveragerc

MatteGary commented:
Is everything all right with this PR? Can it be approved?

MartinHjelmare commented:
Can be merged when a paragraph has been added in the PR description about the breaking change for the release notes.

MartinHjelmare commented:
Please open an issue and link to the PR instead of reporting bugs here.

MatteGary commented:
@kennedyshead @MartinHjelmare I may have an idea on why it happend, I'll wait for the issue to try to fix it.

kennedyshead commented:
Sorry @MartinHjelmare I did not realize you merged it
